lib/graphql/type/composite_type.ex

```diff
@@ -0,0 +1,36 @@
+
+defmodule GraphQL.Type.CompositeType do
+  @moduledoc ~S"""
+  Provides *get_field* and *get_fields* accessors for composite types.
+  This abstacts over the *fields* key being a map or a function that returns a map.
+  """
+  alias GraphQL.Type.{ObjectType, Union, Interface, Input}
+
+  def has_field?(%Interface{fields: fields}, field_name), do: !!_get_field(fields, field_name)
+  def has_field?(%ObjectType{fields: fields}, field_name), do: !!_get_field(fields, field_name)
+  def has_field?(%Union{types: types}, field_name), do: !!_get_field(types, field_name)
+  def has_field?(%Input{fields: fields}, field_name), do: !!_get_field(fields, field_name)
+
+  def get_fields(%Interface{fields: fields}), do: _get_fields(fields)
+  def get_fields(%ObjectType{fields: fields}), do: _get_fields(fields)
+  def get_fields(%Union{types: types}), do: _get_fields(types)
+  def get_fields(%Input{fields: fields}), do: _get_fields(fields)
+
+  def get_field(%Union{types: types}, field_name), do: _get_field(types, field_name)
+  def get_field(%Interface{fields: fields}, field_name), do: _get_field(fields, field_name)
+  def get_field(%ObjectType{fields: fields}, field_name), do: _get_field(fields, field_name)
+  def get_field(%Input{fields: fields}, field_name), do: _get_field(fields, field_name)
+
+  defp _get_fields(fields) when is_function(fields), do: fields.()
```

Comment:
the convention I've seen is having the private function called `do_XXXXX`. The `_` always means that a variable isn't used and this would be confusing.